lab4d/nnutils/nerf.py

```diff
@@ -728,26 +772,33 @@ def importance_sampling(
 
         return xyz_cam, dir_cam, deltas, depth
 
-    @train_only_fields
-    def compute_jacobian(self, xyz, inst_id, field2cam):
+    def compute_jacobian(self, xyz, xyz_cam, dir_cam, field2cam, frame_id, inst_id, samples_dict):
         """Compute eikonal and normal fields from Jacobian of SDF
 
         Args:
-            xyz: (M,N,D,3) Canonical field coordinates
+            xyz: (M,N,D,3) Points along rays in object canonical space. Only for training
+            xyz_cam: (M,N,D,3) Points along rays in camera space. Only for rendering
+            dir_cam: (M,N,D,3) Ray directions in camera space. Only for rendering
+            field2cam: (M,SE(3)) Object-to-camera SE(3) transform. Only for rendering
+            frame_id: (M,) Frame id to query articulations, or None to use all frames.
+                Only for rendering
             inst_id: (M,) Instance id. If None, compute for the average instance
-            field2cam: (M,SE(3)) Object-to-camera SE(3) transform
+            samples_dict (Dict): Time-dependent bone articulations. Only for rendering. Keys:
+                "rest_articulation": ((M,B,4), (M,B,4)) and
+                "t_articulation": ((M,B,4), (M,B,4))
         Returns:
-            jacob_dict (Dict): Jacobian fields. Keys: "eikonal" (M,N,D,1) and
-                "normal" (M,N,D,3)
+            jacob_dict (Dict): Jacobian fields. Keys: "eikonal" (M,N,D,1). Only when
+                rendering, "normal" (M,N,D,3)
         """
         jacob_dict = {}
-        jacob_dict["eikonal"], normal = self.compute_eikonal(xyz, inst_id=inst_id)
-        normal = quaternion_apply(
```

Comment:
This is no longer necessary, as it is accounted for by differentiating through `backward_warp` within `compute_normal`. The previous method also did not account for changes in the normal due to delta skinning or articulations.